crates/pathfinder/src/consensus/inner/p2p_task_tests.rs

```diff
@@ -0,0 +1,1402 @@
+                assert_eq!(
+                    signed_proposal.proposal.height, expected_height,
+                    "Proposal height should match"
+                );
+                assert_eq!(
+                    signed_proposal.proposal.value.0, expected_commitment,
+                    "Proposal commitment should match"
+                );
+            }
+            _ => panic!("Expected Proposal command"),
+        }
+    }
+
+    /// Helper: Verify proposal parts are persisted in database
+    fn verify_proposal_parts_persisted(
+        consensus_storage: &pathfinder_storage::Storage,
+        height: u64,
+        round: u32,
+        validator_address: &ContractAddress, // Query with validator address (receiver)
+        expected_min_parts: usize,
+    ) {
+        let mut db_conn = consensus_storage.connection().unwrap();
+        let db_tx = db_conn.transaction().unwrap();
+        // seems like foreign_proposal_parts queries by validator_address to get
+        // proposals from foreign validators (proposals where proposer !=
+        // validator)
+        let parts = foreign_proposal_parts(&db_tx, height, round, validator_address)
+            .unwrap()
+            .unwrap_or_default();
+
+        // ----- Debug output for proposal parts -----
+        #[cfg(debug_assertions)]
+        {
+            eprintln!(
+                "Found {} proposal parts in database for height {} round {} (querying with \
+                 validator {})",
+                parts.len(),
+                height,
+                round,
+                validator_address.0
+            );
+            for (i, part) in parts.iter().enumerate() {
+                eprintln!("  Part {}: {:?}", i, std::mem::discriminant(part));
+            }
+        }
+        // ----- End debug output for proposal parts -----
+
+        assert!(
+            parts.len() >= expected_min_parts,
+            "Expected at least {} proposal parts, got {}",
+            expected_min_parts,
+            parts.len()
+        );
+    }
+
+    /// Helper: Verify transaction count from persisted proposal parts
+    fn verify_transaction_count(
+        consensus_storage: &pathfinder_storage::Storage,
+        height: u64,
+        round: u32,
+        validator_address: &ContractAddress,
+        expected_count: usize,
+    ) {
+        let mut db_conn = consensus_storage.connection().unwrap();
+        let db_tx = db_conn.transaction().unwrap();
+        let parts = foreign_proposal_parts(&db_tx, height, round, validator_address)
+            .unwrap()
+            .unwrap_or_default();
+
+        // Count transactions from all TransactionBatch parts
+        let mut total_transactions = 0;
+        for part in &parts {
+            if let ProposalPart::TransactionBatch(transactions) = part {
+                total_transactions += transactions.len();
+            }
+        }
+
+        assert_eq!(
+            total_transactions, expected_count,
```

Comment:
I guess this verification makes sense if the fake data is diverse enough. Otherwise we could have false positives.

Reply:
Ok, my comment actually makes more sense only if a test refers to a scenario which spans more than 1 height.